### ISSUE TITLE
LDAP configuration failed (/api/v2/settings/all/ 500 error)

### ISSUE THREAD

**Issue by cello86** [ansible/awx#1704]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 - UI

##### SUMMARY
We tried to configure the LDAP endpoint from web UI but after the change if we click to save and we switch to another tab, the old configurations were missing. We noticed an HTTP 500 error on call /api/v2/settings/all/. We added the stackt race generated during the save action.

##### ENVIRONMENT
* AWX version: 1.0.5.13
* AWX install method: openshift 
* Ansible version: 2.4.2.0
* Operating System: CentOS 7
* Web Browser: Firefox and Chrome

##### STEPS TO REPRODUCE
- Edit the configuration from settings panel and click to save

##### EXPECTED RESULTS
We would keep persistent the LDAP configuration for the external users

##### ACTUAL RESULTS
The configurations becomes blank when i change the tab selection and i return to the previous one.

##### ADDITIONAL INFORMATION
This is the error reported during the issue:

```
10.130.2.1 - - [29/Mar/2018:06:53:59 +0000] "GET /api/v2/settings/all/ HTTP/1.1" 200 9706 "https://awx-web-svc-test-dev.devapps.test.it/" "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.108 Safari/537.36"
[pid: 35|app: 0|req: 91/535] 10.130.2.1 () {50 vars in 2655 bytes} [Thu Mar 29 06:53:58 2018] GET /api/v2/settings/all/ => generated 9706 bytes in 288 msecs (HTTP/1.1 200) 8 headers in 260 bytes (1 switches on core 0)
RESULT 2
OKREADY
2018-03-29 06:54:04,888 ERROR    django.request Internal Server Error: /api/v2/settings/all/
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/usr/lib/python2.7/site-packages/awx/wsgi.py", line 65, in _legacy_get_response
    return super(AWXWSGIHandler, self)._legacy_get_response(request)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
    response = self._get_response(request)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/utils/decorators.py", line 185, in inner
    return func(*args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/generics.py", line 284, in dispatch
    return super(APIView, self).dispatch(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 494, in dispatch
    response = self.handle_exception(exc)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 454, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 491, in dispatch
    response = handler(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/generics.py", line 260, in patch
    return self.partial_update(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/generics.py", line 745, in partial_update
    return super(RetrieveUpdateAPIView, self).partial_update(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/mixins.py", line 84, in partial_update
    return self.update(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/generics.py", line 741, in update
    return super(RetrieveUpdateAPIView, self).update(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/mixins.py", line 69, in update
    serializer.is_valid(raise_exception=True)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/serializers.py", line 236, in is_valid
    self._validated_data = self.run_validation(self.initial_data)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/serializers.py", line 435, in run_validation
    value = self.to_internal_value(data)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/serializers.py", line 465, in to_internal_value
    validated_value = field.run_validation(primitive_value)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/fields.py", line 523, in run_validation
    value = self.to_internal_value(data)
  File "/usr/lib/python2.7/site-packages/awx/conf/serializers.py", line 57, in to_internal_value
    obj = super(SettingFieldMixin, self).to_internal_value(value)
  File "/usr/lib/python2.7/site-packages/awx/sso/fields.py", line 401, in to_internal_value
    return cls(**params_sanitized)
TypeError: __init__() takes at least 2 arguments (2 given)
2018-03-29 06:54:04,888 ERROR    django.request Internal Server Error: /api/v2/settings/all/
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/usr/lib/python2.7/site-packages/awx/wsgi.py", line 65, in _legacy_get_response
    return super(AWXWSGIHandler, self)._legacy_get_response(request)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
    response = self._get_response(request)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/utils/decorators.py", line 185, in inner
    return func(*args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/generics.py", line 284, in dispatch
    return super(APIView, self).dispatch(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 494, in dispatch
    response = self.handle_exception(exc)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 454, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 491, in dispatch
    response = handler(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/generics.py", line 260, in patch
    return self.partial_update(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/generics.py", line 745, in partial_update
    return super(RetrieveUpdateAPIView, self).partial_update(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/mixins.py", line 84, in partial_update
    return self.update(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/generics.py", line 741, in update
    return super(RetrieveUpdateAPIView, self).update(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/mixins.py", line 69, in update
    serializer.is_valid(raise_exception=True)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/serializers.py", line 236, in is_valid
    self._validated_data = self.run_validation(self.initial_data)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/serializers.py", line 435, in run_validation
    value = self.to_internal_value(data)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/serializers.py", line 465, in to_internal_value
    validated_value = field.run_validation(primitive_value)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/fields.py", line 523, in run_validation
    value = self.to_internal_value(data)
  File "/usr/lib/python2.7/site-packages/awx/conf/serializers.py", line 57, in to_internal_value
    obj = super(SettingFieldMixin, self).to_internal_value(value)
  File "/usr/lib/python2.7/site-packages/awx/sso/fields.py", line 401, in to_internal_value
    return cls(**params_sanitized)
TypeError: __init__() takes at least 2 arguments (2 given)
[pid: 36|app: 0|req: 155/536] 10.130.2.1 () {56 vars in 2830 bytes} [Thu Mar 29 06:54:04 2018] PATCH /api/v2/settings/all/ => generated 41 bytes in 325 msecs (HTTP/1.1 500) 5 headers in 177 bytes (1 switches on core 0)
```

**Comment by Riccardo1976** [comment]
The issue reads a lot like one i ran into a while ago. Can you check https://github.com/ansible/awx/issues/1040 and see if you have the same results ?

**Comment by RasimKalimullin** [comment]
LDAP USER DN TEMPLATE  is a required field, without it I get a 500 error.  But if this field is filled in, authentication does not work. I had to remove the value of AUTH_LDAP_USER_DN_TEMPLATE from the database, then everything worked

**Comment by grahamn-gr** [comment]
I've been using AWX 1.0.1.8 and LDAP has been working fine, but when using the same settings on 1.0.5.18 I get the same issues as reported in this ticket.

Filling in "LDAP USER DN TEMPLATE" field didn't work for me. 

In the end I followed what was in #1040 and this resolved my issues. 
The key difference being that I had to change the "LDAP GROUP TYPE" from "MemberDNGroupType" to "GroupOfNamesType" and change the "LDAP GROUP SEARCH" from "(objectClass=group)" to "(objectClass=groupOfNames)". After making these changes LDAP works fine.

**Comment by cello86** [comment]
@Riccardo1976 I tested the same configuration reported into #1040 and it works fine.

**Comment by sonjz** [comment]
I had a similar issue, but I think its just user error and erroneous syntax not being reported on the front-end.

When I had curly braces instead of round braces, I got a javascript error on save (which also produced the web log error above)

On LDAP USER DN TEMPLATE:
`sAMAccountName=%{user}s,OU=Group,DC=company,DC=local`

Same problem for LDAP USER SEARCH:
```
[
"OU=Technology,OU=Toronto YTZ,OU=GoogleApps,DC=company,DC=local",
"SCOPE_SUBTREE",
"(sAMAccountName=%{user}s)"
]
```

In both scenarios, when I changed `%{user}s` to `%(user)s` the javascript error ceased, and my weblog reported a 200 status for the Save, and I confirmed LDAP settings were saved.

On save, it would be helpful to report user error when syntax validation fails.
